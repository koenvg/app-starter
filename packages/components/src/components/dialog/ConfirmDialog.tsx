import { Button } from "../button";
import React from "react";
import { Dialog, DialogActions, DialogTitle, DialogContent, Typography } from "@material-ui/core";

export interface Props {
  onFinish: (confirmed: boolean) => void;
  onConfirm?: () => Promise<unknown>;
  title: string;
  open: boolean;
}

export const ConfirmDialog: React.SFC<Props> = ({ onFinish, title, open, onConfirm }) => {
  const [loading, setLoading] = React.useState(false);
  const [error, setError] = React.useState<Error>();
  const handleClose = React.useCallback(() => onFinish(false), [onFinish]);

  const handleOk = React.useCallback(async () => {
    setLoading(true);
    try {
      if (onConfirm) {
        await onConfirm();
      }
      onFinish(true);
    } catch (e) {
      setLoading(false);
      setError(e);
    }
  }, [onConfirm, onFinish]);

  return (
    <Dialog open={open} onClose={handleClose}>
      <DialogTitle>{title}</DialogTitle>
      <DialogContent>{error && <Typography color="error">{error.message}</Typography>}</DialogContent>
      <DialogActions>
        <Button onClick={handleClose}>Close</Button>
        <Button onClick={handleOk} color="primary" autoFocus loading={loading}>
          Ok
        </Button>
      </DialogActions>
    </Dialog>
  );
};

ConfirmDialog.displayName = "ConfirmDialog";
