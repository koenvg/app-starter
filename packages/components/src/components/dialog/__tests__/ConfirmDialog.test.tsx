import { ConfirmDialog, Props } from "../ConfirmDialog";
import React from "react";
import { render, fireEvent, wait, waitForElement } from "@testing-library/react";

describe("ConfirmDialog", () => {
  const defaultProps: Props = {
    onFinish: jest.fn(),
    open: true,
    title: "Some title",
  };
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render the dialog", () => {
    render(<ConfirmDialog {...defaultProps} />);
    expect(document.body).toMatchSnapshot();
  });

  it("should say true when confirming the dialog", () => {
    const { getByText } = render(<ConfirmDialog {...defaultProps} />);
    fireEvent.click(getByText("Ok"));
    expect(defaultProps.onFinish).toHaveBeenCalledWith(true);
  });

  it("should say false when cancelling the dialog", () => {
    const { getByText } = render(<ConfirmDialog {...defaultProps} />);
    fireEvent.click(getByText("Close"));
    expect(defaultProps.onFinish).toHaveBeenCalledWith(false);
  });

  it("should show the loading state when the onConfirm prop is passed", async () => {
    const { getByText } = render(<ConfirmDialog {...defaultProps} onConfirm={jest.fn().mockResolvedValue({})} />);
    await wait(() => fireEvent.click(getByText("Ok")));
    expect(document.body).toMatchSnapshot();
  });

  it("should show error if the onConfirm throws", async () => {
    const error = new Error("Epic failure");
    const { getByText, queryByText } = render(
      <ConfirmDialog {...defaultProps} onConfirm={jest.fn().mockRejectedValue(error)} />,
    );
    await wait(() => fireEvent.click(getByText("Ok")));
    await waitForElement(() => queryByText("Ok"));
    expect(queryByText(error.message)).not.toBeNull();
  });
});
