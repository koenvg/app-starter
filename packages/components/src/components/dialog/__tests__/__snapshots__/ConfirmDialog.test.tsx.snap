// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialog should render the dialog 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            Some title
          </h2>
        </div>
        <div
          class="MuiDialogContent-root"
        />
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Close
            </span>
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-focusVisible Mui-focusVisible"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Ok
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`ConfirmDialog should show the loading state when the onConfirm prop is passed 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            Some title
          </h2>
        </div>
        <div
          class="MuiDialogContent-root"
        />
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Close
            </span>
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <div
                class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                role="progressbar"
                style="width: 25px; height: 25px;"
              >
                <svg
                  class="MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </div>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
